### PR TITLE
fix constructor arguments verifications

### PR DESCRIPTION
**Pull request by ayrat555**
- Some constructor arguments have multiple constructors arguments appended at the end
- Older smart contracts don't have a whisper code 

## Changelog
- fix nested constructor arguments
- fix verification of older smart contracts with constructor args

### CODE REVIEW

File: apps/explorer/lib/explorer/smart_contract/verifier/constructor_arguments.ex [hunk 1]
```diff
@@ -5,31 +5,52 @@ defmodule Explorer.SmartContract.Verifier.ConstructorArguments do
...
+    creation_code
+    |> String.split(contract_code)
+    |> List.last()
     |> Kernel.==(arguments_data)
   end
 
-  defp extract_constructor_arguments(<<>>), do: ""
+  defp extract_constructor_arguments(code, passed_constructor_arguments) do
+    case code do
+      "a165627a7a72305820" <> <<_::binary-size(64)>> <> "0029" <> constructor_arguments ->
```
Comment (vbaranov) [inline apps/explorer/lib/explorer/smart_contract/verifier/constructor_arguments.ex]: Could you drop a comment here to which Solidity/EVM version does this case relate?

File: apps/explorer/lib/explorer/smart_contract/verifier/constructor_arguments.ex [hunk 2]
```diff
@@ -5,31 +5,52 @@ defmodule Explorer.SmartContract.Verifier.ConstructorArguments do
...
+        if passed_constructor_arguments == constructor_arguments do
+          true
+        else
+          extract_constructor_arguments(constructor_arguments, passed_constructor_arguments)
+        end
 
-  defp extract_constructor_arguments("a165627a7a72305820" <> <<_::binary-size(64)>> <> "0029" <> constructor_arguments) do
-    constructor_arguments
-  end
+      "a265627a7a72305820" <>
```
Comment (vbaranov) [inline apps/explorer/lib/explorer/smart_contract/verifier/constructor_arguments.ex]: the same as above

File: apps/explorer/lib/explorer/smart_contract/verifier/constructor_arguments.ex [hunk 3]
```diff
@@ -5,31 +5,52 @@ defmodule Explorer.SmartContract.Verifier.ConstructorArguments do
 
   alias Explorer.Chain
 
-  def verify(address_hash, arguments_data) do
-    arguments_data = String.replace(arguments_data, "0x", "")
+  def verify(address_hash, contract_code, arguments_data) do
+    arguments_data = arguments_data |> String.trim_trailing() |> String.trim_leading() |> String.replace("0x", "")
```
Comment (pasqu4le) [inline apps/explorer/lib/explorer/smart_contract/verifier/constructor_arguments.ex]: AFAIK `"0x"` is to be removed from the beginning of the string only.
If this is the case, I think changing
` |> String.trim_trailing() |> String.trim_leading() |> String.replace("0x", "")`
to
` |> String.trim() |> String.trim_leading("0x")`
would make for a cleaner solution.

File: apps/explorer/lib/explorer/smart_contract/verifier/constructor_arguments.ex (hunk 1 above)
Comment (ayrat555) [reply apps/explorer/lib/explorer/smart_contract/verifier/constructor_arguments.ex]: It's not defined in the documentation from which version it does in the docs (https://github.com/ethereum/solidity/issues/4021)
I added a comment to the specific version which uses it

File: apps/explorer/lib/explorer/smart_contract/verifier/constructor_arguments.ex (hunk 2 above)
Comment (ayrat555) [reply apps/explorer/lib/explorer/smart_contract/verifier/constructor_arguments.ex]: done

File: apps/explorer/lib/explorer/smart_contract/verifier/constructor_arguments.ex (hunk 3 above)
Comment (ayrat555) [reply apps/explorer/lib/explorer/smart_contract/verifier/constructor_arguments.ex]: done

**Pull request merged**